packages/token-generation-readmodel-checker/src/utils/utils.ts

```diff
@@ -0,0 +1,1242 @@
+    logger.error(
+      `Client states are not equal:
+  platform-states entry: ${
+    platformStatesClientEntry
+      ? JSON.stringify(
+          ComparisonPlatformStatesClientEntry.parse(platformStatesClientEntry)
+        )
+      : platformStatesClientEntry
+  }
+  client read-model: ${JSON.stringify(ComparisonClient.parse(client))}`
+    );
+  }
+
+  return {
+    isPlatformStatesClientCorrect,
+    data:
+      !isPlatformStatesClientCorrect && platformStatesClientEntry
+        ? {
+            PK: platformStatesClientEntry.PK,
+            clientKind: platformStatesClientEntry.clientKind,
+            clientConsumerId: platformStatesClientEntry.clientConsumerId,
+            clientPurposesIds: platformStatesClientEntry.clientPurposesIds,
+          }
+        : undefined,
+  };
+}
+
+// eslint-disable-next-line sonarjs/cognitive-complexity
+function validateTokenGenerationStates({
+  tokenGenStatesEntries,
+  client,
+  purposesMap,
+  consumerIdEserviceIdMap,
+  eserviceIdDescriptorIdMap,
+  logger,
+}: {
+  tokenGenStatesEntries: TokenGenerationStatesGenericClient[] | undefined;
+  client: Client;
+  purposesMap: Map<PurposeId, Purpose>;
+  consumerIdEserviceIdMap: Map<GSIPKConsumerIdEServiceId, Agreement>;
+  eserviceIdDescriptorIdMap: Map<GSIPKEServiceIdDescriptorId, EService>;
+  logger: Logger;
+}): {
+  isTokenGenerationStatesClientCorrect: boolean;
+  data: ComparisonTokenGenStatesGenericClient[] | undefined;
+} {
+  if (!tokenGenStatesEntries || tokenGenStatesEntries.length === 0) {
+    if (client.keys.length === 0) {
+      return {
+        isTokenGenerationStatesClientCorrect: true,
+        data: undefined,
+      };
+    }
+
+    console.log(
+      `Client ${client.id} has ${client.keys.length} ${
+        client.keys.length > 1 ? "keys" : "key"
+      } but zero token-generation-states entries`
+    );
+    logger.error(
+      `Client ${client.id} has ${client.keys.length} ${
+        client.keys.length > 1 ? "keys" : "key"
+      } but zero token-generation-states entries`
+    );
+    return {
+      isTokenGenerationStatesClientCorrect: false,
+      data: undefined,
+    };
+  }
+
+  const tokenGenStatesEntriesCount =
+    client.purposes.length > 0
+      ? client.keys.length * client.purposes.length
+      : client.keys.length;
+  const wrongTokenGenStatesEntries = tokenGenStatesEntries.reduce<
+    ComparisonTokenGenStatesGenericClient[]
+  >(
+    (acc, e) =>
+      match(e)
+        // eslint-disable-next-line complexity
+        .with({ clientKind: clientKindTokenGenStates.consumer }, (e) => {
+          if (client.purposes.length !== 0) {
+            // TokenGenerationStatesConsumerClient with CLIENTKIDPURPOSE PK
+            const purposeId = getPurposeIdFromTokenGenStatesPK(e.PK);
+            const purpose = purposeId ? purposesMap.get(purposeId) : undefined;
+
+            if (!purpose) {
+              if (TokenGenerationStatesClientKidPK.safeParse(e.PK).success) {
```

Comment:
It's preferable to verify that a value is the expect X, instead of checking that it is not the unexpected Y.
As an example, if the PK is neither a ClientKid or a ClientKidPurpose the current check would succeed

```suggestion
              if (TokenGenerationStatesClientKidPurposePK.safeParse(e.PK).success) {
              <switch the conditions>
```